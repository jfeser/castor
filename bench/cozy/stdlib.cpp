#include "../../etc/date.c"

using namespace std;

float int_to_float(int x) {
  return static_cast<float>(x);
}

bool streq(string s1, string s2) {
  return s1 == s2;
}

int parse_date(string d) {
  int year, mon, day;
  sscanf(d.c_str(), "%4d-%2d-%2d", &year, &mon, &day);
  return days_from_civil(year, mon, day);
}

int to_month(int d) {
  int year, mon, day;
  civil_from_days(d, &year, &mon, &day);
  return mon;
}


template<typename T> void sink(T const& t) {
  volatile T sinkhole = t;
}
