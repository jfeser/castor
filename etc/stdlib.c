#include <stdio.h>
#include <time.h>

long strpos(char* s1, long l1, char* s2, long l2) {
  if (l1 < l2) { return 0; }
  for (int i = 0; i < l1; i++) {
    int eq_cnt = 0;
    for (int j = 0; j < l2; j++) {
      eq_cnt += s1[i + j] == s2[j];
    }
    if (eq_cnt == l2) {return i + 1;}
  }
  return 0;
}

long extract_year(long date) {
  time_t time = date * 24 * 60 * 60;
  struct tm btime;
  gmtime_r(&time, &btime);
  // tm_year is the number of years since 1900
  return btime.tm_year + 1900;
}

long extract_month(long date) {
  time_t time = date * 24 * 60 * 60;
  struct tm btime;
  gmtime_r(&time, &btime);
  // tm_mon ranges from 0 to 11
  return btime.tm_mon + 1;
}

long extract_day(long date) {
  time_t time = date * 24 * 60 * 60;
  struct tm btime;
  gmtime_r(&time, &btime);
  // tm_day is an ordinal unlike everything else
  return btime.tm_mday;
}

