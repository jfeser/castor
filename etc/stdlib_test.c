#include <stdio.h>
#include "stdlib.c"

void strpos_test(char *s1, long l1, char* s2, long l2) {
  printf("strpos(%s, %ld, %s, %ld) = %ld\n",
         s1, l1, s2, l2, strpos(s1, l1, s2, l2));
}

int main() {
  strpos_test("testing", 7, "in", 2);
  strpos_test("tested", 6, "in", 2);
}
