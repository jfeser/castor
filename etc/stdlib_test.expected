strpos(testing, 7, in, 2) = 5
strpos(tested, 6, in, 2) = 0
